Stop generating nondet pointees for function pointers

xbauch opened the pull request:
when building the entry point. Because that leads to declaring function-type
objects and then assigning values to them. Rather we should simply initialise
the function pointer with NONDET(fptr_type) and let the function-pointer removal
find the candidates.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e5d5c9faf2d577e3163a114d451aef206c1ee5d2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120202139

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 68c9554a3d1bd658a20969cb09775fa568a0fa75).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144957668

hannes-steffenhagen-diffblue reviewed (approved):
This comes up a lot, maybe we should have a utility function like `is_pointer_to_object` or something of the sort?

thk123 commented on regression/cbmc/Function_Pointer_Init/main.c:
```diff
@@ -0,0 +1,10 @@
+
+void foo(other_function_type other_function)
+{
+  // returning from the function call is unreachable -> the following assertion
```
:question: This comment doesn't make sense wrt to the test checking for verification failed?

xbauch replied:
Extended.

thk123 commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -48,7 +48,7 @@ void symbol_factoryt::gen_nondet_init(
   }
 
-  if(type.id()==ID_pointer)
+  if(type.id() == ID_pointer && to_pointer_type(type).subtype().id() != ID_code)
```
:question: What happens with pointers to code then? Just random number? 
Read the PR description - that is worth a comment here explaining why we don't need to initialize it here

xbauch replied:
Now the symbol is explicitly assigned a nondet (of the function-ptr type).

thk123 reviewed (commented):
:no_entry_sign: Please add a test for when there _is_ a function of the right type, to check it is selected. And two functions of the right type, to check it picks between them (as I believe the new behaviour should / old behaviour did not).

xbauch commented:
>  Please add a test for when there _is_ a function of the right type, to check it is selected. And two functions of the right type, to check it picks between them (as I believe the new behaviour should / old behaviour did not).

I've added the tests. Note that existing functions will only be considered as candidates if dereferenced somewhere in the code.

petr-bauch committed:
Stop generating nondet pointees for function pointers

when building the entry point. Because that leads to declaring function-type
objects and then assigning values to them. Rather we should simply initialise
the function pointer with NONDET(fptr_type) and let the function-pointer removal
find the candidates.

thk123 commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -54,6 +54,13 @@ void symbol_factoryt::gen_nondet_init(
 
+    if(subtype.id() == ID_code)
+    {
+      assignments.add(
```
:pick: I'd still like a comment to the effect 
```
// leave as nondet_ptr to allow `remove_function_pointers`
// to replace the pointer. 
```

xbauch replied:
Added.

thk123 reviewed (approved):
LGTM, thanks for the extra tests. 
~This just needs code owner approval @tautschnig @kroening~

thk123 commented:
currently failing CI, but looks transient, have restarted the job

petr-bauch committed:
Add comment to the change in symbol-factory

will be squashed.

thk123 commented:
This is now passing CI and just needs code owner approval @tautschnig @kroening
